src/lib/scheduler/node_queue_scheduler.cpp

```diff
@@ -76,22 +82,46 @@ void NodeQueueScheduler::wait_for_all_tasks() {
       // concurrent queues when many tiny tasks have been scheduled (see MergeSort scheduler test). When this assert is
       // triggered in other situations, there have probably been new tasks added after wait_for_all_tasks() was called.
       Assert(queue_check_runs < 1'000, "Queue is not empty but all registered tasks have already been processed.");
+
+      // Explicitly signal workers to work on remaining jobs.
+      // TODO(Martin): still needed?
+      // queue->signal(1);
+
       std::this_thread::sleep_for(std::chrono::milliseconds(1));
       ++queue_check_runs;
     }
   }
 }
 
 void NodeQueueScheduler::finish() {
+  // Lock finish() to ensure that the shutdown tasks are not sent twice.
+  auto lock = std::lock_guard<std::mutex>{_finish_mutex};
+
+  if (!_active) {
+    return;
+  }
+
   wait_for_all_tasks();
 
-  // All queues SHOULD be empty by now
-  if (HYRISE_DEBUG) {
-    for (auto& queue : _queues) {
-      Assert(queue->empty(), "NodeQueueScheduler bug: Queue wasn't empty even though all tasks finished");
+  auto workers_shut_down = std::atomic_uint64_t{_worker_count};
```

Comment:
This is not the number of workers that were shut down, but the number of active workers, right (starts from #workers and is decremented, rahter than starting with 0 and being incremented)? Also, `_workers.size()`?